### PR TITLE
#49: LocalizedException SHOULD only be thrown in View

### PR DESCRIPTION
**Pull request by larsroettig**
see: #49

### CODE REVIEW

File: Magento/ruleset.xml [hunk 1]
```diff
@@ -195,6 +195,10 @@
         <severity>7</severity>
         <type>warning</type>
     </rule>
+    <rule ref="Magento.Exceptions.LocalizedThrow">
+        <severity>7</severity>
```
Comment (lenaorobei) [inline Magento/ruleset.xml]: Please change the severity to `8`. This rule looks like:
> Magento specific code issue.

Comment (lenaorobei) [reply Magento/ruleset.xml]: Sorry for nit-picking, but could you please add rules in alphabetical order? So they will be ordered by severity then by name.

File: Magento/Sniffs/Exceptions/LocalizedThrowSniff.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Sniffs\Exceptions;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+
+/**
+ * Detects possible direct throws of Exceptions.
```
Comment (paliarush) [inline Magento/Sniffs/Exceptions/LocalizedThrowSniff.php]: Please update comment

File: Magento/Sniffs/Exceptions/LocalizedThrowSniff.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ */
+class LocalizedThrowSniff implements Sniff
+{
+
+    protected $presentationLayerKeys = ['Controller', 'Block', 'ViewModel'];
+
+    /**
+     * String representation of warning.
+     */
+    protected $warningMessage = 'LocalizedException SHOULD only be thrown in the Presentation layer.';
```
Comment (paliarush) [inline Magento/Sniffs/Exceptions/LocalizedThrowSniff.php]: It looks like tech guidelines must be updated to allow LocalizedException in Service Layer as well. See #5.17 in https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html

Please do not merge this PR until tech guidelines are updated.

**Comment by larsroettig** [comment]
@paliarush can you give me feedback than i change the implementation

**Comment by larsroettig** [comment]
Close because of Impelementation does not solve it able with the new requirements with tokens  

See:
https://github.com/magento/devdocs/pull/3871

**Pull request closed without merging**